presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java

```diff
@@ -14,86 +14,196 @@
 package com.facebook.presto.cost;
 
 import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static java.lang.Double.NEGATIVE_INFINITY;
+import static java.lang.Double.NaN;
+import static java.lang.Double.POSITIVE_INFINITY;
+import static java.lang.Double.isNaN;
+import static java.lang.Double.max;
+import static java.lang.Double.min;
 import static java.util.stream.Stream.concat;
 
 public class PlanNodeStatsEstimateMath
 {
     private PlanNodeStatsEstimateMath()
     {}
 
-    @FunctionalInterface
-    private interface RangeSubtractionStrategy
+    public static PlanNodeStatsEstimate subtractStats(PlanNodeStatsEstimate superset, PlanNodeStatsEstimate subset)
     {
-        StatisticRange range(StatisticRange leftRange, StatisticRange rightRange);
+        if (superset.isOutputRowCountUnknown() || subset.isOutputRowCountUnknown()) {
+            return PlanNodeStatsEstimate.unknown();
+        }
+
+        double supersetRowCount = superset.getOutputRowCount();
+        double subsetRowCount = subset.getOutputRowCount();
+        double outputRowCount = supersetRowCount - subsetRowCount;
+        verify(outputRowCount >= 0, "outputRowCount must be greater than or equal to zero: %s", outputRowCount);
```

Comment:
we shouldn't fail here as stats are not perfect, we should just cap to 0

Reply:
That's was one of the reason why i wanted to have this inlined. When estimating `NOT` statement we know for sure that the negated stats shouldn't be bigger than the input stats, as the negated stats is a strict subset of the input. Having explicit assertions improved readability and makes it easier to understand what the code does.

Reply:
> When estimating NOT statement we know for sure that the negated stats shouldn't be bigger than the input stats, as the negated stats is a strict subset of the input

Double math is not perfect. It could be that `subset` rows equals to `superset + 0.0000001` rows. Therefore it cannot be assumed that stats will always be within constraints envelope. A lot of your recent refactors do just that (they make calculator more forgiving on case of bogus stats).  It seems we should be consistent here.

Reply:
> Double math is not perfect. It could be that subset rows equals to superset + 0.0000001 rows.

I think it makes sense. Let me change it.

> Why inlining is required?

The idea was that for the NOT estimate all the assertions always true, when for OR estimate (or in some other place where this method could be used) the assertions may start failing. But since we are going to replace assertion with capping - it is no longer the case.